src/workerd/api/pyodide/pyodide.h

```diff
@@ -166,25 +167,61 @@ class PyodideMetadataReader: public jsg::Object {
   }
 };
 
+// Similar to PackagesTarReader, but reads from a dynamic buffer rather than a buffer linked into
+// the binary
+class SmallPackagesTarReader: public jsg::Object {
+  kj::ArrayPtr<const kj::byte> source;
+
+public:
+  SmallPackagesTarReader(kj::ArrayPtr<const kj::byte> src): source(src) {};
+
+  int read(jsg::Lock& js, int offset, kj::Array<kj::byte> buf);
+
+  JSG_RESOURCE_TYPE(SmallPackagesTarReader) {
+    JSG_METHOD(read);
+  }
+};
+
 struct MemorySnapshotResult {
   kj::Array<kj::byte> snapshot;
   kj::Array<kj::String> importedModulesList;
   JSG_STRUCT(snapshot, importedModulesList);
 };
 
+// Before a Pyodide isolate starts up, all its packages begin loading. This struct provides an
+// interface for the Pyodide startup code to request promises for each package that was loaded.
+class PackagePromiseMap {
+  // This implementation is complex because we're not allowed to store kj::Promises on an object
+  // stored on the V8 heap. Ideally, we'd like to simply store a map from package paths to Promises.
+  // Instead, we store a CrossThreadWaitList and get a new Promise from the waitlist as needed.
+  kj::HashMap<kj::String, kj::Own<SmallPackagesTarReader>> fetchedPackages;
+  kj::HashMap<kj::String, CrossThreadWaitList> waitlists;
+
+public:
+  void insert(kj::String path, kj::Promise<kj::Own<SmallPackagesTarReader>> promise);
+
+  // Note it is explicitly not allowed to call getPromise more than once on the same package
+  // This is because allowing this causes ownership of the SmallPackagesTarReader to be muddied.
```

Comment:
What happens when we do this accidentally? Will we get a nice error? If not can you add some code to ensure we do?